druid/src/command.rs

```diff
@@ -15,19 +15,69 @@
 //! Custom commands.
 
 use std::any::Any;
-use std::sync::{Arc, Mutex};
+use std::{
+    marker::PhantomData,
+    sync::{Arc, Mutex},
+};
 
 use crate::{WidgetId, WindowId};
 
+/// An untyped identifier for a `Selector`.
+pub type SelectorSymbol = &'static str;
+
+/// An identifier for a particular command.
+///
+/// This should be a unique string identifier. Certain `Selector`s are defined
+/// by druid, and have special meaning to the framework; these are listed in the
+/// [`druid::commands`] module.
+///
+/// [`druid::commands`]: commands/index.html
+#[derive(Debug, PartialEq, Eq)]
+pub struct Selector<T = ()>(SelectorSymbol, PhantomData<*const T>);
+
+// This has do be done explicitly, to avoid the Copy bound on `T`.
+// See https://doc.rust-lang.org/std/marker/trait.Copy.html#how-can-i-implement-copy .
+impl<T> Copy for Selector<T> {}
+impl<T> Clone for Selector<T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
 /// An identifier for a particular command.
 ///
 /// This should be a unique string identifier. Certain `Selector`s are defined
 /// by druid, and have special meaning to the framework; these are listed in the
 /// [`druid::commands`] module.
 ///
 /// [`druid::commands`]: commands/index.html
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub struct Selector(&'static str);
+#[derive(Debug, PartialEq, Eq)]
+pub struct OneShotSelector<T>(SelectorSymbol, PhantomData<*const T>);
+
+// This has do be done explicitly, to avoid the Copy bound on `T`.
+// See https://doc.rust-lang.org/std/marker/trait.Copy.html#how-can-i-implement-copy .
+impl<T> Copy for OneShotSelector<T> {}
+impl<T> Clone for OneShotSelector<T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+pub trait AnySelector {
```

Comment:
I would try and get rid of this in favour of using some std trait, like maybe `AsRef`?

Reply:
Yes, if `AnySelector` is only used for the symbol, than that could make more sens.
I used a custom trait with the intend to merge `Command::new` and `Command::once`.
If we don't do that, `AsRef` could be fine, even though I like that `AnySelector` makes it clear that it expects a selector.

Reply:
The `AsRef` implementation isn't too confusing I think if we add some documentation that it takes a `Selector` / `SelectorOnce`.

```rust
impl<T> AsRef<SelectorSymbol> for Selector<T> {
    fn as_ref(&self) -> &SelectorSymbol {
        &self.0
    }
}

impl Command {
    pub fn is(&self, selector: impl AsRef<SelectorSymbol>) -> bool {
        self.selector == *selector.as_ref()
    }
}
```